### PR TITLE
[presign] Get Gnosis Safe transaction state and update ethereumTransactionHash

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

This PR includes some new logic in the `FinalizeTxUpdater` logic to:
* 1. Periodically check the state of the gnosis safe transaction
* 2. Update the state with the safe info
* 3. Update the txHash if available
* 4. Obtains the receipt of the Ethereum transaction when it's available
* 5. Finalize the transaction, the same way normal transactions are finalized

image

# Not included
The UX is still not perfect. There's a few details missing, for example if you click in the transaction shown in the side panel, it will take you to a non existing transaction. This is because, for now it just shows you the Gnosis Safe tx hash. I plan to address this in the future. @biocom This relates to create the 2 links: safe tx, etherscan

# To Test
This PR should make approvals/wraps/unwraps work with Gnosis Safe.

You might experience a small delay from when the transaction is mined and when it shows as finalized in CowSwap, however this delay should be a few seconds at most.

For dev testers, you can check the REDUX actions, how periodiacally is updating with the latest Gnosis Safe Tx info, until we have a mined transaction, then we would invoke the tx finalization.

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @anxolin , cool, I was able to get successful statuses for Wrap and Approve transactions. 
But!
1. Statuses are not updated when a transaction is failed: I receive red toast message about failed transaction in Safe, but failed transaction is not displayed in the transaction log.. As a result, it hangs in the Cowswap app:
safe failed
failed transaction
approving hang

2. I'm not able to create an order where token to Buy = ETH
image
image


Sell/Buy orders work fine.

**Review by alfetopito — COMMENTED** [review]
Testing with single owner Safe:

## Approval
Worked, expect for:
1. No notification when tx detected, I thought we had that

After awhile tx changed from `Open` to `Approved` :+1: 

## Swap
Working fine

## (Un)wrapping
Not working, still same error
screenshot_2021-09-24_10-34-27

**Comment by alfetopito** [comment]
>     2. I'm not able to create an order where token to Buy = ETH
>        image
>        image

That is a limitation of the contract. ETH cannot be sent to Smart Contract wallets.
There was a gh-issue to better handle that in the UI but I can't find it right now :/

File: src/custom/state/enhancedTransactions/hooks/index.ts [hunk 1]
```diff
@@ -1,34 +1,37 @@
 import { useCallback, useMemo } from 'react'
 import { useAppDispatch } from 'state/hooks'
 
-import { useActiveWeb3React } from 'hooks/web3'
 import { addTransaction, AddTransactionParams } from '../actions'
 import { EnhancedTransactionDetails, HashType } from '../reducer'
 import { useAllTransactions } from 'state/enhancedTransactions/hooks'
+import { useWalletInfo } from '@src/custom/hooks/useWalletInfo'
```
Comment (alfetopito) [inline src/custom/state/enhancedTransactions/hooks/index.ts:7]: Path?

```suggestion
import { useWalletInfo } from 'hooks/useWalletInfo'
```

**Comment by elena-zh** [comment]
> That is a limitation of the contract. ETH cannot be sent to Smart Contract wallets.
> There was a gh-issue to better handle that in the UI but I can't find it right now :/

As an option, can we disable ETH in the token list when connected to Safe? 
Or, at least, show a warning that this token in not supported with current connection?

**Comment by biocom** [comment]
@elena-zh And I think next to the warning a suggestion saying 'pick WETH instead'. But that will get crowded perhaps.

**Comment by elena-zh** [comment]
@biocom , might be a good option. But need to see it on UI...

**Comment by anxolin** [comment]
Merging to consolidate all the waterfalls. Reviews are still welcomed, the idea is:
- Capture the comments of pending issues in: https://docs.google.com/document/d/1_I0yPgocBqbKGnFLC1L3nvlS3nMXZcNLOPhvR8YAxi4/edit#   
- Aggregate all pending PRs into develop
- Solve all the issues :) 
- Finish the Gnosis Safe / Presign tasks: https://github.com/gnosis/cowswap/issues/1305    
- Do testing once all is fixed and finished. Note that these functionalities required too many PRs to implement in iterations. For QA probably is too many, this is why I want to close a release next week so we can test it all merged together.

**Pull request merged**